mlir/lib/Transforms/NormalizeMemRefs.cpp

```diff
@@ -0,0 +1,221 @@
+
+#include "PassDetail.h"
+#include "mlir/Analysis/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/IR/Function.h"
+#include "mlir/IR/FunctionSupport.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/Utils.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/Format.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "normalize-memrefs"
+
+using namespace mlir;
+
+namespace {
+
+/// All interprocedural memrefs with non-trivial layout maps are converted to
+/// ones with trivial identity layout ones.
+
+// Input :-
+// #tile = affine_map<(i) -> (i floordiv 4, i mod 4)>
+// func @matmul(%A: memref<16xf64, #tile>, %B: index, %C: memref<16xf64>) ->
+// (memref<16xf64, #tile>) {
+//   affine.for %arg3 = 0 to 16 {
+//         %a = affine.load %A[%arg3] : memref<16xf64, #tile>
+//         %p = mulf %a, %a : f64
+//         affine.store %p, %A[%arg3] : memref<16xf64, #tile>
+//   }
+//   %c = alloc() : memref<16xf64, #tile>
+//   %d = affine.load %c[0] : memref<16xf64, #tile>
+//   return %A: memref<16xf64, #tile>
+// }
+
+// Output :-
+// module {
+//   func @matmul(%arg0: memref<4x4xf64>, %arg1: index, %arg2: memref<16xf64>)
+//   -> memref<4x4xf64> {
+//     affine.for %arg3 = 0 to 16 {
+//       %2 = affine.load %arg0[%arg3 floordiv 4, %arg3 mod 4] : memref<4x4xf64>
+//       %3 = mulf %2, %2 : f64
+//       affine.store %3, %arg0[%arg3 floordiv 4, %arg3 mod 4] : memref<4x4xf64>
+//     }
+//     %0 = alloc() : memref<16xf64, #map0>
+//     %1 = affine.load %0[0] : memref<16xf64, #map0>
+//     return %arg0 : memref<4x4xf64>
+//   }
+// }
+
+struct NormalizeMemRefs : public NormalizeMemRefsBase<NormalizeMemRefs> {
+  void runOnOperation() override;
+};
+
+} // end anonymous namespace
+
+std::unique_ptr<OperationPass<ModuleOp>> mlir::createNormalizeMemRefsPass() {
+  return std::make_unique<NormalizeMemRefs>();
+}
+
+void NormalizeMemRefs::runOnOperation() {
+  ModuleOp moduleOp = getOperation();
+
+  moduleOp.walk([&](FuncOp funcOp) {
+    OpBuilder b(funcOp);
+
+    FunctionType ft = funcOp.getType();
+    SmallVector<Type, 8> argTypes;
+    SmallVector<Type, 4> resultTypes;
+
+    // Populating results with function's initial result (type) as
+    // this will be used to modify and set function's signature later.
+    for (unsigned retIndex : llvm::seq<unsigned>(0, funcOp.getNumResults())) {
+      resultTypes.push_back(ft.getResult(retIndex));
+    }
+
+    // Walk over each argument of a function to perform memref normalization (if
+    // any).
+    for (unsigned argIndex : llvm::seq<unsigned>(0, funcOp.getNumArguments())) {
+      MemRefType memrefType;
+      Type argType = funcOp.getArgument(argIndex).getType();
+      // Check whether argument is of MemRef type.
+      if (!(memrefType = argType.dyn_cast<MemRefType>())) {
```

Comment:
You can move the assignment outside the conditional on the declaration line, and then just do `if (!memrefType)` - would be simpler and same number of lines.